src/server/utils/validation/index.js

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020 American Express Travel Related Services Company, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import Joi from '@hapi/joi';
+
+import { provideStateConfigSchema, validateStateConfigSchema } from './stateConfig';
+import { providedExternalSchema, requiredExternalSchema } from './externals';
+import { safeRequestSchema } from './safeRequest';
+import { semverExtension } from './semver';
+import { pwaSchema } from './pwa';
+
+export const rootModuleSchema = Joi.object().keys({
+  providedExternals: providedExternalSchema,
+  provideStateConfig: provideStateConfigSchema,
+  csp: Joi.string().required().messages({
+    'any.required': '"csp" must provide a valid content security policy in the Root module',
+  }),
+  corsOrigins: Joi.array().items(
+    Joi.alternatives().try(
+      Joi.object().instance(RegExp),
+      Joi.string().hostname(),
+      Joi.string().pattern(/(?:https?)?[a-z.-]/)
+    )
+  ),
+  configureRequestLog: Joi.function(),
+  extendSafeRequestRestrictedAttributes: safeRequestSchema,
+  pwa: pwaSchema,
+  createSsrFetch: Joi.function(),
+  eventLoopDelayThreshold: Joi.number().allow(Infinity),
+  appCompatibility: semverExtension.semver().valid(),
```

Comment:
Doesn't this come in as 

```json
"app": {
  "compatibility : "x.x.x"
}
```

I forget where the conversion to `appCompatibility` comes into play.

Reply:
Yes! `@americanexpress/one-app-bundler` actually sets this on the modules meta data. When we come to validate it, we can access it from the module and make sure it's compatible with the running One App instance.